src/charm.py

```diff
@@ -379,6 +412,100 @@ def _on_website_relation_joined(self, event: RelationEvent) -> None:
+        if not data:
+            return
+
+        if not self._state.is_ready():
+            event.defer()
+            return
+
+        available_addresses = self._get_available_pro_airgapped_server_addresses(event.relation)
+        if len(available_addresses) == 0:
+            logger.error("expected at least one pro-airgapped-server address, but so far none provided by the relation")
+            return
+
+        logger.debug("found %d available addresses for pro-airgapped-server", len(available_addresses))
+
+        # Only change the address already stored in the state, if the old one is
+        # None or no longer available. This is to avoid unnecessary service restarts.
+        current_address = self._state.pro_airgapped_address
+        if not current_address or current_address not in available_addresses:
+            address = available_addresses.pop()
+            logger.info("changing pro-airgapped-server address to %s", address)
+            self._state.pro_airgapped_address = address
+            self._config_changed(event)
+
+    def _on_pro_airgapped_server_relation_departed(self, event: RelationDepartedEvent):
+        """Handle pro-airgapped-server relation-departed event."""
+        if not self.model.unit.is_leader():
+            return
+        if not self._state.is_ready():
+            event.defer()
+            return
+
+        if len(event.relation.units) == 0:
+            # There is no pro-airgapped-server unit related to this charm, so we should opt out
+            # of using airgapped contracts.
+            logger.warning("falling back to default behavior since all pro-airgapped-server units quit the relation")
+            del self._state.pro_airgapped_address
+            self._config_changed(event)
+            return
+
+        # There might be other units of the pro-airgapped-server in the relation, so we should
+        # use the URL from them. Note that this is a very rare scenario, because the airgapped
+        # contracts service is usually deployed with one unit. But we nevertheless we have to
+        # handle this scenario.
+        available_addresses = self._get_available_pro_airgapped_server_addresses(event.relation)
+        if len(available_addresses) > 0:
+            current_address = self._state.pro_airgapped_address
+            if current_address not in available_addresses:
+                address = available_addresses.pop()
+                logger.info("changing pro-airgapped-server address to %s", address)
+                self._state.pro_airgapped_address = address
+                self._config_changed(event)
+            return
+
+        # We couldn't find a valid address to the pro-airgapped-server, so we have to fallback to default.
+        logger.warning("falling back to default behavior since no valid address to pro-airgapped-server is available")
+        del self._state.pro_airgapped_address
+        self._config_changed(event)
```

Comment:
Ah I didn't realise that one unit departing would also trigger this event. Good job handling this.

Reply:
Sorry I had suggested a change and then realised that we needed all this logic, that's why I highlighted so many lines.